### PR TITLE
ccip - fix ocr3 keyring adapter, transmitter, evm keyring lib

### PR DESCRIPTION
**Pull request by dimkouv**
Fixing the wrong ocr3 keyring adapter for ccip v1.6 by using ocr3 keyring and removing extraHash (unused) from reportContext, making report context `[2][32]byte`

NOTE: This update introduces correct sequence numbers to the contracts and you might see reverts in testnet reports since offRamp was storing/caching the wrong sequence number so far.

### CODE REVIEW

File: core/services/keystore/keys/ocr2key/evm_keyring_test.go [hunk 1]
```diff
@@ -96,3 +97,66 @@ func TestEVMKeyring_Marshalling(t *testing.T) {
 	// Invalid seed size should error
 	assert.Error(t, kr2.Unmarshal([]byte{0x01}))
 }
+
+func TestRawReportContext3(t *testing.T) {
```
Comment (0xAustinWang) [inline core/services/keystore/keys/ocr2key/evm_keyring_test.go:101]: thanks for the test

File: core/capabilities/ccip/ocrimpls/keyring.go [hunk 2]
```diff
@@ -6,16 +6,26 @@ import (
...
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/ocr2key"
 )
 
+// OCR3SignerVerifierExtra is an extension of OCR3SignerVerifier that
+// also exposes the public key and max signature length which are required by the ocr3Keyring adapter.
+type OCR3SignerVerifierExtra interface {
+	ocr2key.OCR3SignerVerifier
+	PublicKey() types.OnchainPublicKey
+	MaxSignatureLength() int
```
Comment (0xAustinWang) [inline core/capabilities/ccip/ocrimpls/keyring.go:17]: how was this working before?

**Review by samsondav — DISMISSED** [review]

**Review by mateusz-sekara — DISMISSED** [review]

File: core/capabilities/ccip/ocrimpls/keyring.go (hunk 2 above)
Comment (dimkouv) [reply core/capabilities/ccip/ocrimpls/keyring.go:17]: it was not working correctly, was providing wrong values to the contract

**Review by jmank88 — DISMISSED** [review]

**Review by 0xAustinWang — DISMISSED** [review]

File: core/capabilities/ccip/ocrimpls/contract_transmitter.go [hunk 3]
```diff
@@ -13,18 +13,23 @@ import (
...
+func ToCommitCalldata(rawReportCtx [2][32]byte, report []byte, rs, ss [][32]byte, vs [32]byte) any {
 	// Note that the name of the struct field is very important, since the encoder used
 	// by the chainwriter uses mapstructure, which will use the struct field name to map
 	// to the argument name in the function call.
 	// If, for whatever reason, we want to change the field name, make sure to add a `mapstructure:"<arg_name>"` tag
 	// for that field.
+
+	// WARNING: Be careful if you change the data types.
+	// Using a different type e.g. `type Foo [32]byte` instead of `[32]byte`
+	// will trigger undefined chainWriter behavior, e.g. transactions submitted with wrong arguments.
```
Comment (makramkd) [inline core/capabilities/ccip/ocrimpls/contract_transmitter.go:30]: 😢

File: core/services/keystore/keys/ocr2key/evm_keyring.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ func (ekr *evmKeyring) reportToSigData3(digest types.ConfigDigest, seqNr uint64,
 
 func RawReportContext3(digest types.ConfigDigest, seqNr uint64) [2][32]byte {
 	seqNrBytes := [32]byte{}
-	binary.BigEndian.PutUint64(seqNrBytes[:], seqNr)
+	binary.BigEndian.PutUint64(seqNrBytes[24:], seqNr)
```
Comment (makramkd) [inline core/services/keystore/keys/ocr2key/evm_keyring.go:63]: Might be worth asking Mercury devs whether they rely on this behavior, if they do, might be better to make this injectable into the keyring

**Review by RensR — DISMISSED** [review]

File: core/services/keystore/keys/ocr2key/evm_keyring.go (hunk 4 above)
Comment (RensR) [reply core/services/keystore/keys/ocr2key/evm_keyring.go:63]:
```
  /// @inheritdoc IVerifier
  function verify(
    bytes calldata signedReport,
    address sender
  ) external override returns (bytes memory verifierResponse) {
    if (msg.sender != i_verifierProxyAddr) revert AccessForbidden();
    (
      bytes32[3] memory reportContext,
      bytes memory reportData,
      bytes32[] memory rs,
      bytes32[] memory ss,
      bytes32 rawVs
    ) = abi.decode(signedReport, (bytes32[3], bytes, bytes32[], bytes32[], bytes32));

    // The feed ID is the first 32 bytes of the report data.
    bytes32 feedId = bytes32(reportData);

    VerifierState storage feedVerifierState = s_feedVerifierStates[feedId];

    // If the feed has been deactivated, do not verify the report
    if (feedVerifierState.isDeactivated) {
      revert InactiveFeed(feedId);
    }

    // reportContext consists of:
    // reportContext[0]: ConfigDigest
    // reportContext[1]: 27 byte padding, 4-byte epoch and 1-byte round
    // reportContext[2]: ExtraHash
    bytes32 configDigest = reportContext[0];
    Config storage s_config = feedVerifierState.s_verificationDataConfigs[configDigest];

    _validateReport(feedId, configDigest, rs, ss, s_config);
    _updateEpoch(reportContext, feedVerifierState);

    bytes32 hashedReport = keccak256(reportData);

    _verifySignatures(hashedReport, reportContext, rs, ss, rawVs, s_config);
    emit ReportVerified(feedId, sender);

    return reportData;
  }
```


Mercury code, they are on OCR2 it appears. They do run OCR3 offchain right? Just have not updated the keyring like we did in this PR?

Comment (makramkd) [reply core/services/keystore/keys/ocr2key/evm_keyring.go:63]: Yeah, it certainly seems like it, though I recall Mercury devs added this initially, so wanted 2 be sure

**Review by makramkd — APPROVED** [review]

**Review by asoliman92 — APPROVED** [review]

**Review by asoliman92 — CHANGES_REQUESTED** [review]
We need to update function `execute` as well

**Review by asoliman92 — APPROVED** [review]

**Review by mateusz-sekara — APPROVED** [review]

**Review by RensR — APPROVED** [review]

**Pull request merged**